String.cxx

```diff
@@ -0,0 +1,184 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <algorithm>
+#include <cctype>
+#include <charconv>
+#include <cstring>
+#include <iostream>
+#include <regex>
+#include <string>
+
+enum StringBaseConv {
+    BIN = 2,
+    DEC = 10,
+    HEX = 16,
+};
+
+enum class byte : unsigned char {};
+
+class String {
+    std::string m_u;
+
+    String(std::string u) : m_u{std::move(u)} {}
+
+  public:
+    String() noexcept = default;
+    String(const String&) = default;
+    String(String&&) noexcept = default;
+    String& operator=(const String&) = default;
+    String& operator=(String&&) = default;
+
+    template <std::size_t N> inline String(const char (&charr)[N]) : m_u{charr, N} {}
+    inline String(const char* cstr) : m_u{cstr} {}
+    inline explicit String(char c) : m_u{1, c} {}
+    template <class T, class = std::enable_if_t<std::is_integral<T>::value>> inline String(T val, StringBaseConv base) {
+        constexpr static auto base2len_upperbound = [](StringBaseConv base) /*constexpr*/ {
+            switch (base) {
+            case BIN:
+                return 65;
+            case DEC:
+
+            case HEX:
+                return 17;
+            default:
+                throw;
+            }
+        };
+        m_u.resize(base2len_upperbound(base));
+        const auto res = std::to_chars(&*m_u.begin(), &*m_u.rbegin(), val, +base);
+        if (static_cast<int>(res.ec))
+            throw;
+        m_u.resize(std::strlen(m_u.c_str()));
+    }
+
+     // template <class T, class = std::enable_if_t<std::is_floating_point<T>::value>>
+    // String(T val, int precision); // unimplemented
+
+    [[nodiscard]] inline auto c_str() const noexcept { return m_u.c_str(); }
+    [[nodiscard]] inline auto length() const noexcept { return m_u.length(); }
+    [[nodiscard]] inline auto charAt(unsigned idx) const noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto& charAt(unsigned idx) noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto operator[](unsigned idx) const noexcept { return m_u[idx]; }
+    [[nodiscard]] inline auto& operator[](unsigned idx) noexcept { return m_u[idx]; }
+
+    template <class T>[[nodiscard]] inline bool concat(const T& v) {
+        m_u += String(v).m_u;
+        return true;
+    }
+
+    inline int compareTo(const String& s) const noexcept { return std::memcmp(m_u.c_str(), s.m_u.c_str(), (std::min)(s.m_u.size(), m_u.size())); }
+
+    inline bool startsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(m_u.c_str(), s.m_u.c_str(), s.m_u.size()) == 0;
+    }
+
+    inline bool endsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(&*m_u.rbegin() - s.m_u.size(), s.m_u.c_str(), s.m_u.size()) == 0;
```

Comment:
std::memcmp(m_u.data() + m_u.size() - s.m_u.size() - 1, s.m_u.c_str(), s.m_u.size())
Would also be a little more readable, although on this im not sure but I don't like the &*m_u.rbegin().

Reply:
Guess I'll polyfill `<compare>` and `std::lexicographical_compare_three_way`